aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,339 @@
+
+.. _add-header-attribute:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/AddHeaderAttribute.cs
+  :language: c#
+  :emphasize-lines: 5,16
+
+Attributes allow filters to accept arguments, as shown in the example above. You would add this attribute to a controller or action method and specify the name and value of the HTTP header you wished to add to the response:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/SampleController.cs
+  :language: c#
+  :emphasize-lines: 1
+  :lines: 6-12,25
+  :dedent: 4
+
+The result of the ``Index`` action is shown below - the response headers are displayed on the bottom right.
+
+.. image:: filters/_static/add-header.png
+
+The available built-in attribute-based filters are:
+	- `ActionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ActionFilterAttribute/index.html>`_
+	- `AuthorizationFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/AuthorizationFilterAttribute/index.html>`_
+	- `ExceptionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ExceptionFilterAttribute/index.html>`_
+	- `ResultFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ResultFilterAttribute/index.html>`_
+
+Dependency Injection
+^^^^^^^^^^^^^^^^^^^^
+Filters are not directly returned from :doc:`dependency injection </fundamentals/dependency-injection>` (DI). Filters that are implemented as attributes and added directly to controller classes or action methods cannot have constructor dependencies provided by DI. This is because attributes must have their constructor parameters supplied where they are applied. This is a limitation of how attributes work.
+
+However, if your filters have dependencies you need to access from DI, there are several supported approaches. You can apply your filter to a class or action method using
+
+	- The `ServiceFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/ServiceFilterAttribute/index.html>`_ 
+	- The `TypeFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/TypeFilterAttribute/index.html>`_ attribute
+	- `IFilterFactory <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/IFilterFactory/index.html>`__ implemented on your attribute
+	- Access to the services you require from the context parameter provided to your filter's On\ *Stage* methods.
```

Comment:
Using `IServiceProvider` should be your last resort when you need access to services. We provide first class mechanisms for filters. Please don't mention it.